File: src/cz/encircled/eplayer/service/VLCMediaService.java
package cz.encircled.eplayer.service;

import com.sun.jna.Memory;
import com.sun.jna.Native;
import com.sun.jna.NativeLibrary;
import cz.encircled.eplayer.common.Constants;
import cz.encircled.eplayer.model.MediaType;
import cz.encircled.eplayer.service.event.Event;
import cz.encircled.eplayer.service.event.EventObserver;
import cz.encircled.eplayer.service.gui.ViewService;
import cz.encircled.eplayer.util.GuiUtil;
import cz.encircled.eplayer.util.Localizations;
import cz.encircled.eplayer.util.LocalizedMessages;
import cz.encircled.eplayer.view.AppView;
import javafx.application.Platform;
import javafx.scene.image.PixelFormat;
import javafx.scene.image.PixelWriter;
import javafx.scene.image.WritablePixelFormat;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import uk.co.caprica.vlcj.binding.LibVlc;
import uk.co.caprica.vlcj.component.DirectMediaPlayerComponent;
import uk.co.caprica.vlcj.player.MediaPlayer;
import uk.co.caprica.vlcj.player.MediaPlayerEventAdapter;
import uk.co.caprica.vlcj.player.direct.BufferFormat;
import uk.co.caprica.vlcj.player.direct.BufferFormatCallback;
import uk.co.caprica.vlcj.player.direct.DirectMediaPlayer;
import uk.co.caprica.vlcj.player.direct.format.RV32BufferFormat;
import uk.co.caprica.vlcj.runtime.RuntimeUtil;

import javax.annotation.PostConstruct;
import javax.annotation.Resource;
import javax.swing.*;
import java.awt.*;
import java.nio.ByteBuffer;
import java.util.concurrent.CountDownLatch;

import static cz.encircled.eplayer.util.LocalizedMessages.ERROR_TITLE;
import static cz.encircled.eplayer.util.LocalizedMessages.MSG_VLC_LIBS_FAIL;

/**
 * Created by Administrator on 9.6.2014.
 */
@Resource
public class VLCMediaService implements MediaService {

    private static final Logger log = LogManager.getLogger();
    public static final String X64 = "64";

    @Resource
    private CacheService cacheService;

    @Resource
    private GuiUtil guiUtil;

    @Resource
    private Localizations localizations;

    @Resource
    private EventObserver eventObserver;

    private long currentTime;

    private DirectMediaPlayerComponent mediaPlayerComponent;

    private DirectMediaPlayer player;

    private WritablePixelFormat<ByteBuffer> pixelFormat;

    @Resource
    private AppView appView;

    /**
     * Implementation of a direct rendering media player component that renders
     * the video to a JavaFX canvas.
     */
    private class TestMediaPlayerComponent extends DirectMediaPlayerComponent {

        @Override
        public void display(DirectMediaPlayer mediaPlayer, Memory[] nativeBuffers, BufferFormat bufferFormat) {
            Memory nativeBuffer = nativeBuffers[0];
            ByteBuffer byteBuffer = nativeBuffer.getByteBuffer(0, nativeBuffer.size());
            PixelWriter pixelWriter = appView.getPixelWriter();
            Platform.runLater(new Runnable() {
                @Override
                public void run() {
                    pixelWriter.setPixels(0, 0, bufferFormat.getWidth(), bufferFormat.getHeight(), pixelFormat, byteBuffer, bufferFormat.getPitches()[0]);
                }
            });

        }

        public TestMediaPlayerComponent() {
            super(new TestBufferFormatCallback());
        }
    }

    private class TestBufferFormatCallback implements BufferFormatCallback {

        @Override
        public BufferFormat getBufferFormat(int sourceWidth, int sourceHeight) {
            /* TODO event
            Platform.runLater(new Runnable() {
                @Override
                public void run() {
                    canvas.setWidth(width);
                    canvas.setHeight(height);
                    stage.setWidth(width);
                    stage.setHeight(height);
                }
            });          */

            return new RV32BufferFormat(sourceWidth, sourceHeight);
        }
    }

    @Resource
    private ViewService viewService;

    @Nullable
    private String current;

    public static final String VLC_LIB_PATH_64 = "vlc-2.1.5_x64";

    public static final String VLC_LIB_PATH = "vlc-2.1.5";

    @Override
    public void releasePlayer(){
//        if(player != null){
//            player.stop();
//            player.release();
//            current = null;
//            currentTime = 0;
//        }
    }

    @Override
    public void updateCurrentMediaInCache(){
        if(current != null)
            cacheService.updateEntry(current.hashCode(), currentTime);
    }

    private void play(@NotNull String path, long time){

        CountDownLatch playerCountDown = new CountDownLatch(1);
        viewService.showPlayer(playerCountDown);
        try {
            playerCountDown.await();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        log.debug("play: {}", EventQueue.isDispatchThread());
        if(!path.equals(current)){
            log.debug("Path {} is new", path);
            current = path;
            player.prepareMedia(path);
        }
        player.start();
        player.setTime(Math.min(time, player.getLength()));
        eventObserver.fire(Event.SubtitlesUpdated, player.getSpuDescriptions());
        eventObserver.fire(Event.AudioTracksUpdated, player.getAudioDescriptions());
        eventObserver.fire(Event.PlayStart);
    }

    @Override
    public void play(@NotNull String path){
        play(cacheService.createIfAbsent(path));
    }

    @Override
    public void play(@NotNull MediaType p){
        play(p.getPath(), p.getTime());
    }

    @Override
    public void setSubtitles(int id) {
        player.setSpu(id);
    }

    @Override
    public void setAudioTrack(int trackId) {
        player.setAudioTrack(trackId);
    }

    @Override
    public int getMediaLength() {
        return (int) player.getLength();
    }

    @Override
    public int getCurrentTime() {
        return (int) player.getTime();
    }

    @Override
    public int getVolume() {
        return player.getVolume();
    }

    @Override
    public void setVolume(int value) {
        player.setVolume(value);
    }

    @Override
    public void setTime(long value) {
        player.setTime(value);
    }

    @Override
    public void start() {
        player.start();
    }

    @Override
    public void pause() {
        if (player.isPlaying())
             player.pause();
    }

    @Override
    public void togglePlayer() {
        if(player != null && current != null)
            player.pause();
    }

    @Override
    public void stop(){
        if(player != null){
            current = null;
            player.stop();
            viewService.enableSubtitlesMenu(false);
        }
    }

    @PostConstruct
    private void init() {
        long start = System.currentTimeMillis();
        initializeLibs();
        pixelFormat = PixelFormat.getByteBgraInstance();
        log.trace("VLCMediaService init start");
        try {
            mediaPlayerComponent = new TestMediaPlayerComponent();

            player = mediaPlayerComponent.getMediaPlayer();
//            player.setEnableMouseInputHandling(false);
//            player.setEnableKeyInputHandling(false);
                                    /*
            mediaPlayerComponent.getVideoSurface().addMouseListener(new MouseAdapter() { // TODO move

                @Override
                public void mouseClicked(@NotNull MouseEvent e) {
                    if(e.getClickCount() == Constants.ONE){
                        togglePlayer();
                    } else {
                        togglePlayer();
                    }
                }
            });            */

            player.addMediaPlayerEventListener(new MediaPlayerEventAdapter() {
                @Override
                public void finished(MediaPlayer mediaPlayer) {
                    if (current != null) {
                        cacheService.updateEntry(current.hashCode(), Constants.ZERO_LONG);
                    }
                    current = null;
                    stop();
                    viewService.showQuickNavi();
                }

                @Override
                public void timeChanged(MediaPlayer mediaPlayer, long newTime) {
                    currentTime = newTime;
                    eventObserver.fire(Event.MediaTimeChange, newTime);
                }

                @Override
                public void error(MediaPlayer mediaPlayer) {
                    log.error("Failed to open media {} ", current);
                    guiUtil.showMessage(LocalizedMessages.FILE_OPEN_FAILED, LocalizedMessages.ERROR_TITLE, JOptionPane.ERROR_MESSAGE);
                    if(current != null) {
                        cacheService.deleteEntry(current.hashCode());
                        current = null;
                    }
                    viewService.showQuickNavi();
                }

            });

        } catch(Exception e){
            log.error("Player initialization failed", e);
            guiUtil.showMessage("VLC library not found", "Error title", JOptionPane.ERROR_MESSAGE);
        } catch (NoClassDefFoundError re){
            log.error("Player initialization failed", re);
        }
        log.trace("VLCMediaService init complete in {} ms", System.currentTimeMillis() - start);
    }

    private void initializeLibs(){
        long start = System.currentTimeMillis();
        log.trace("Initialize VLC libs");
        String vlcLibPath = System.getProperty("sun.arch.data.model").equals(X64) ? VLC_LIB_PATH_64 : VLC_LIB_PATH;
        NativeLibrary.addSearchPath(RuntimeUtil.getLibVlcLibraryName(), vlcLibPath);
        try {
            Native.loadLibrary(RuntimeUtil.getLibVlcLibraryName(), LibVlc.class);
            log.trace("VLCLib successfully initialized in {} ms", System.currentTimeMillis() - start);
        } catch(UnsatisfiedLinkError e){
            guiUtil.showMessage(localizations.get(MSG_VLC_LIBS_FAIL), localizations.get(ERROR_TITLE), JOptionPane.ERROR_MESSAGE);
            log.error("Failed to load vlc libs from specified path {}", vlcLibPath);
            // TODO exit?
        }
    }

}

File: src/main/java/cz/encircled/eplayer/service/VLCMediaService.kt
package cz.encircled.eplayer.service

import cz.encircled.eplayer.core.ApplicationCore
import cz.encircled.eplayer.model.GenericTrackDescription
import cz.encircled.eplayer.model.MediaSeries
import cz.encircled.eplayer.model.PlayableMedia
import cz.encircled.eplayer.service.event.Event
import cz.encircled.eplayer.service.event.MediaCharacteristic
import cz.encircled.eplayer.util.TimeMeasure.measure
import cz.encircled.eplayer.view.AppView
import cz.encircled.eplayer.view.UiUtil.inNormalThread
import org.apache.logging.log4j.LogManager
import uk.co.caprica.vlcj.factory.MediaPlayerFactory
import uk.co.caprica.vlcj.player.base.MediaPlayer
import uk.co.caprica.vlcj.player.component.EmbeddedMediaPlayerComponent
import java.io.File
import java.util.concurrent.CountDownLatch

/**
 * @author Encircled on 9.6.2014.
 */
class VLCMediaService(private val core: ApplicationCore) : MediaService {

    private val log = LogManager.getLogger()
    private val vlcArgs = "--mmdevice-passthrough=2"

    private var current: PlayableMedia? = null

    lateinit var player: MediaPlayer

    override var volume: Int
        get() = player.audio().volume()
        set(value) {
            if (value != player.audio().volume()) {
                player.audio().setVolume(value)
                Event.volumeChanged.fire(value)
            }
        }

    private var subtitle: GenericTrackDescription? = null

    private var audioTrack: GenericTrackDescription? = null

    init {
        Event.audioTrackChanged.listen { change ->
            switchAudioPassThruIfNeeded(change.characteristic)
        }
    }

    override fun currentSubtitle() = subtitle

    override fun setSubtitle(track: GenericTrackDescription, byUser: Boolean) {
        if (track != subtitle) {
            log.debug("Set subtitle track {}", track)
            subtitle = track
            player.subpictures().setTrack(track.id)
            Event.subtitleChanged.fire(MediaCharacteristic(current!!, track, byUser))
        }
    }

    override fun currentAudioTrack() = audioTrack

    override fun setAudioTrack(track: GenericTrackDescription, byUser: Boolean) {
        if (track != audioTrack) {
            log.debug("Set audio track {}", track)
            this.audioTrack = track
            player.audio().setTrack(track.id)
            Event.audioTrackChanged.fire(MediaCharacteristic(current!!, track, byUser))
        }
    }

    override fun currentMedia(): PlayableMedia? = current

    override fun createPlayer(): EmbeddedMediaPlayerComponent {
        releasePlayer()

        val mediaPlayerFactory = MediaPlayerFactory(if (core.settings.audioPassThrough) vlcArgs else "")
        val playerComponent = EmbeddedMediaPlayerComponent(mediaPlayerFactory, null, null, null, null)

        this.player = playerComponent.mediaPlayer()
        this.player.events().addMediaPlayerEventListener(VLCEventListener(core, this))

        return playerComponent
    }

    override fun isPlaying(): Boolean = player.status().isPlaying

    override fun play(path: String) = play(core.cacheService.createIfAbsent(path))

    override fun play(media: PlayableMedia) {
        if (!this::player.isInitialized) return

        core.cacheService.createIfAbsent(media)

        val path = media.mediaFile().path
        log.info("Play {}, start time is {}", path, media.time)

        measure("Show player on UI") {
            val countDownLatch = CountDownLatch(1)
            core.appView.showPlayer(countDownLatch)
            countDownLatch.await()
        }

        subtitle = null
        audioTrack = null
        current = media

        inNormalThread {
            player.media().start(path)

            setTime(media.time.get())
        }

        log.debug("Playing started")
    }

    override fun playNext() {
        val media = current
        if (media is MediaSeries) {
            stop()
            media.toNext()
            play(media)
        }
    }

    override fun playPrevious() {
        val media = current
        if (media is MediaSeries) {
            stop()
            media.toPrev()
            play(media)
        }
    }

    override fun setTimePlus(value: Long) {
        if (current != null) {
            val newTime = player.status().time() + value
            setTime(newTime)
            Event.mediaTimeChange.fire(MediaCharacteristic(current!!, newTime), true)
        }
    }

    override fun setTime(value: Long) {
        if (player.status().isPlayable) {
            player.controls().setTime(value)
        } else {
            // Restart the media from given time, if it has already finished
            if (value > 0) {
                current?.let {
                    play(it)
                    player.controls().setTime(value)
                }
            }
        }
    }

    override fun start() {
        player.controls().start()
    }

    override fun toggle() {
        if (player.status().isPlayable) {
            if (player.status().isPlaying) {
                player.controls().pause()
            } else {
                player.controls().play()
            }
        } else {
            // Restart the media, if it has finished
            current?.let { play(it) }
        }
    }

    override fun pause() {
        if (player.status().isPlaying) player.controls().pause()
    }

    override fun stop() {
        current?.mediaFile()?.let {
            player.snapshots().save(
                File(ApplicationCore.getScreenshotLocation(it)),
                AppView.SCREENSHOT_WIDTH,
                AppView.SCREENSHOT_HEIGHT
            )
        }

        log.debug("Stop player")
        player.controls().stop()
        current = null
    }

    override fun releasePlayer() {
        if (this::player.isInitialized) {
            stop()
            player.release()
        }
    }

    override fun switchAudioPassThruIfNeeded(track: GenericTrackDescription) {
        val isRequired = core.mediaSettingsSuggestions.suggestAudioPassThroughRequired(track)
        if (isRequired && !core.settings.audioPassThrough) {
            core.settings.audioPassThrough(true)
            Event.audioPassThroughChange.fire(true)
        }
    }

}
